lib/cn/csched_sp3.c

```diff
@@ -1931,27 +1936,64 @@ sp3_tree_shape_check(struct sp3 *sp)
                 lsiz = pcap;
                 lclen = cn_ns_clen(&tn->tn_ns) >> 20;
             }
-        } else if (cn_node_isroot(tn)) {
-            if (!rlen_node || len > rlen) {
-                rlen_node = tn;
-                rlen = len;
-            }
-        } else {
-            if (!ilen_node || len > ilen) {
-                ilen_node = tn;
-                ilen = len;
+        }
+        rmlock_runlock(lock);
+
+        if (len > 0)
+            continue;
+
+        /* We currently have no way to delete the rightmost node if it is
+         * empty because the work scheduling and compaction job apparatus
+         * requires at least one kvset in a node.  Until that changes, we
+         * can look for and remove all rightmost empty nodes periodically
+         * as we perform this shape check.
+         */
```

Comment:
Comment is a bit confusing because it says there's no way to delete empty rightmost nodes, yet here we are deleting empty rightmost nodes.

Why is the rightmost node special in this regard?  What about other nodes?

Reply:
I could clarify it, there's no way to submit a job to delete an empty node because all the compaction infrastructure is kvset based.  Similarly, joins do not fit into the existing scheduler/compaction infrastructre because it's all single-node centric.  So, the anchor node of a join is the node with 1 or more kvsets that makes it through sp3_work(), and the node to be merged into the anchor can be empty, but must always be to the left of the anchor node so that we can make the incremental spill logic work correctly when walking the tree (ispill must never interleave across two nodes of a join.

Reply:
Simplest case where this is a problem is a monotonically decreasing time series.  If we do not clean up the rightmost empty nodes then we'll hit max fanout and not be able to create any new nodes.
I suppose we could allow the right node of a join to be empty, but that makes things even more complicated (bad enough as it is!)

Reply:
I'll rework the comment..

Reply:
Reworked, better?

Reply:
Yeah.  All I was looking for was qualifying  "we cannot delete empty nodes" with "as a compaction job".   :)